### PR TITLE
cmake: restructure CMakeLists.txt to make it a single project

### PR DESCRIPTION
**Pull request by arnout**
This will allow to build everything in the regular CMake way. Follow-up
commits will refine things for a better out-of-the box experience.

All cmake files are collected in the top-level cmake directory, which is
the one that gets included in the search path.

Create a top-level CMakeLists.txt that contains the common parts. It
adds the four components as subdirectories. All the pieces of the
top-level CMakeLists.txt files that are used in at least two components
are moved here.

Add a bunch of TODO entries for future improvements.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

### CODE REVIEW

File: framework/tlvf/CMakeLists.txt
Comment (tomereli) [inline framework/tlvf/CMakeLists.txt]: The below install(PROGRAMS ${TLVF_DIR}/tlvf.py DESTINATION host/bin) should also be removed as part of this commit, and same for the target exporting (this goes for all target exports which are only used internally)

**Review by tomereli — APPROVED** [review]

**Comment by arnout** [comment]
At the moment, everything builds on my machine (for Linux), except with dwpal:
1. it can't find libsafec-3.0 while linking;
2. it gives a warning that turns into an error with -Werror - this is probably unrelated, I haven't tested building with dwpal in a while.

I have not tested on RDKB or UGW, of course. I'm also still going to check the rpath values and run the executables.

I propose that @tomereli tests the build on RDKB and UGW and tells me what I have to fix (I'd be surprised if it works :-).

The idea of this PR was to do the very minimum to be able to build with the top-level CMakeLists.txt. However, it turns out that that very minimum is quite a lot (current diffstat is +537-228, 44 files). However, I don't see an easy way to split it up. It would be possible to first move the installation paths around so the changes to the #include's can be done in a separate patch; but that's again a lot of work, and it requires other changes that will be undone later on.

There are a number of changes in this PR that I'm not so sure of, so please consider those carefully:
* Headers are no longer installed, so it is not possible to create another program that communicates with the rest over the local bus, or uses other features (e.g. bpl, bwl) in a different context.
* #include's have changed from e.g. `<beerocks/bwl/mon_wlan_hal.h>` to just `<mon_wlan_hal.h>`, maybe that is less clear (because the component is no longer specified).

So I propose the following approach:
1. @tomereli revies this PR in its current state.
2. @tomereli tests on RDK-B and UGW
3. @arnout addresses the remaining review comments either in this PR or in follow-up PRs, as appropriate.
4. @arnout updates the description so it's appropriate as a commit message
5. @arnout fixes the dwpal issues in follow-up PRs.
6. @arnout does some more cleanup of CMakeLists.txt in follow-up PRs.

Examples of cleanups that should still be done:
* Some stuff gets installed that probably shouldn't.
* Lots of redundant things in lower CMakeLists.txt files, like `cmake_minimum_required(VERSION 2.8)` and `target_include_directories(btl PRIVATE $<TARGET_PROPERTY:bcl,INTERFACE_INCLUDE_DIRECTORIES>)` (this is empty because bcl no longer exports an INTERFACE).
* Unify/simplify the way some things are handled (e.g. there are at least 3 different ways that elpp gets included).
* Move some more common functionality up to top-level.
* Expand the GLOB expressions ([GLOB](https://cmake.org/cmake/help/v3.3/command/file.html?highlight=GLOB) is [evil](https://stackoverflow.com/questions/32411963/why-is-cmake-file-glob-evil)).
* Fix more things so CMakeLists.txt can be used directly in an IDE (allows/simplifies building/debugging for Linux straight from the IDE).
* More things I'm forgetting :-)

**Comment by arnout** [comment]
Another important improvement: fix the Tlvf generator so you don't need to rerun cmake every time a yaml file is changed. Since now everything is built, re-running cmake also means that all object files are rebuilt so it really takes a long time.

File: common/beerocks/bcl/CMakeLists.txt
Comment (tomereli) [inline common/beerocks/bcl/CMakeLists.txt]: comment should be removed - library is no longer importable

File: controller/src/beerocks/bml/CMakeLists.txt
Comment (tomereli) [inline controller/src/beerocks/bml/CMakeLists.txt:40]: I can't comment on the commit message, but you are right - bml must be exported, this is the interface to the controller. So all the removed exports which are needed need to come back.

File: agent/src/beerocks/slave/CMakeLists.txt
Comment (tomereli) [inline agent/src/beerocks/slave/CMakeLists.txt]: I'm confused - how did it work without it? What was the -I path before this change?

File: agent/src/beerocks/monitor/mon_wlan_hal_loader.cpp
Comment (tomereli) [inline agent/src/beerocks/monitor/mon_wlan_hal_loader.cpp:10]: For me, it is more readable having the submodule prefix in the include, we try to keep this convention throughout the code.
We could keep it as `#include <bwl/mon_wlan_hal.h>` and move the public headers from `bwl/common` to `bwl/include/beerocks/bwl` like we do in the framework.

What do you think?

Edit - I see you referred to this in your comment. Bottom line is that I think it is more readable having at least the module name as a prefix, but it's up for debate.

File: common/beerocks/bwl/CMakeLists.txt
Comment (tomereli) [inline common/beerocks/bwl/CMakeLists.txt:109]: Do we still need the INSTALL_INTERFACE here?

File: agent/src/beerocks/slave/CMakeLists.txt
Comment (tomereli) [inline agent/src/beerocks/slave/CMakeLists.txt:34]: The slave is dynamically loading bpl, will it only fix the include path (if it's wrong then how did it work before?) - but will it have other effect?

File: framework/examples/transport/CMakeLists.txt
Comment (tomereli) [inline framework/examples/transport/CMakeLists.txt:1]: It used to work but probably got broken long time ago :)

**Review by tomereli — CHANGES_REQUESTED** [review]
There are 3 major issues which need to be solved before we can merge this -
1. BML has to be exported, it is used by both UGW and RDKB components
2. With this, it is no longer possible to build specific components. It forces the building of framework+ common + controller + agent - this might be needed for the controller node, but definitely not if we build agent only.
We need to have the option to build:
- EasyMesh controller - common + framework + agent + controller
- EasyMesh agent - common + framework + agent

As a matter of fact, it will still be useful to be able to build each component on its own, to make sure we don't add unneeded dependencies.

3. Exported targets
The framework should still be exporting its targets - 
- BPL will be used by an external platform service
- TLVF  and the local bus might be used by dDWRT services outside of the EasyMesh specification to support actions not in the standard (for example - reset all nodes in the network, set up parental control, etc.)
- BTL may need to be exported (along with BCL) if we decide to implement dDWRT services on top of 1905 as vendor specific CMDUs

**Comment by tomereli** [comment]
> Another important improvement: fix the Tlvf generator so you don't need to rerun cmake every time a yaml file is changed. Since now everything is built, re-running cmake also means that all object files are rebuilt so it really takes a long time.

We want to re-run the generator (tlvf.py) when yaml files change. What we don't want is for tlvf.py to re-generate *all* yaml files, including the ones which didn't change. 
If I remember correctly, we had a task for it in our backlog (@LiorAmram please correct me if I'm mistaken) - tlvf.py should first read the autogenerated file if it exists, compare it to what's about to be written, and if they are identical, skip it.

Added a task for this - https://github.com/prplfoundation/prplMesh/issues/10

File: controller/src/beerocks/bml/CMakeLists.txt
Comment (arnout) [reply controller/src/beerocks/bml/CMakeLists.txt:40]: I propose to first remove it all, and then add it again later in a consistent way. Does that work for you?

File: agent/src/beerocks/slave/CMakeLists.txt
Comment (arnout) [reply agent/src/beerocks/slave/CMakeLists.txt]: target_link_libraries automatically pulls in the interface properties (for an external dependency) resp. the public properties (for an internal dependency). Thus, the PRIVATE include_directories property of beerocks slave will be updated with the PUBLIC include_directories property of bwl after this change.

Before this change, bwl headers were installed by the common module, and this module would just find them in the install/include directory that is added explicitly in each CMakeLists.txt file.

File: agent/src/beerocks/monitor/mon_wlan_hal_loader.cpp
Comment (arnout) [reply agent/src/beerocks/monitor/mon_wlan_hal_loader.cpp:10]: I agree, but I would do such cleanups in later commits.

File: common/beerocks/bwl/CMakeLists.txt
Comment (arnout) [reply common/beerocks/bwl/CMakeLists.txt:109]: We don't, but I would do such cleanups in later commits - we need many more like that. It definitely doesn't harm (INTERFACE is only used by export install).

File: framework/examples/transport/CMakeLists.txt
Comment (arnout) [reply framework/examples/transport/CMakeLists.txt:1]: Indeed, I tried on the master branch and it also breaks in the same way.

File: controller/src/beerocks/bml/CMakeLists.txt
Comment (tomereli) [reply controller/src/beerocks/bml/CMakeLists.txt:40]: Works fine as long as we keep it separate (not merge it yet), since I want to sync our internal repos to what we did so far and this will break it, what do you think?

File: agent/src/beerocks/slave/CMakeLists.txt
Comment (tomereli) [reply agent/src/beerocks/slave/CMakeLists.txt]: But what about the fact that the slave uses dlopen to load BWL? Won't this mean also that BWL will be dynamically linked and there would be no need for dlopen?

Comment (arnout) [reply agent/src/beerocks/slave/CMakeLists.txt]: It is true that with this change, the slave will be linked with bwl, while before bwl was only loaded with dlopen().

But I don't see the use of using dlopen() here. It's not really a plugin, because you still need to give a CMake option to say which bwl to build. And even without dlopen, you can at runtime replace the .so with a different .so that exports the same functions, so I don't see the point. dlopen() would only be useful if you could choose at runtime which bwl to use - but I don't see why that is useful.

**Pull request closed without merging**
